kw6/reader.py

```diff
@@ -143,68 +145,80 @@ def __getitem__(self, indices_or_slice):
                 ]
 
         elif isinstance(indices_or_slice, Iterable):
-            positions = [self.position_(index) for index in indices_or_slice]
+            positions = [self.seek_position_(index) for index in indices_or_slice]
 
         else:
             raise TypeError(f"Unindexable type {type(indices_or_slice)}")
 
         return positions
 
-    def position_(self, frame_index: types.FRAME_INDEX):
+    def seek_position_(self, frame_index: types.FRAME_INDEX):
         if frame_index < 0:
-            raise IndexError("Negative indexing not supported.")
-        try:
-            if self.inconsistent_position_bytes:
-                raise IOError
-            self.assumptuous_seek_(frame_index)
-            position = Position.from_stream_(self.stream)
-        except Exception:
-            self.inconsistent_position_bytes = True
-            self.seek_(frame_index)
-            if not self.empty():
-                position = Position.from_stream_(self.stream)
-            else:
-                raise IndexError(frame_index)
+            raise IndexError("Negative indexing not supported")
 
-        if position.header.frame_index != frame_index:
+        if frame_index < self.initial_frame_index:
             raise IndexError(
-                f"Expected to find {frame_index}, "
-                f"but found {position.header.frame_index} instead."
+                f"Frame index {frame_index} is smaller than the first frame index"
             )
 
-        return position
+        attempts = 0
+        position = None
+        step_size_confidence = -1
+        while not self.empty():
+            self.stream.seek(self.closest_stored_byte_position(frame_index))
+            current_frame_index = PositionHeader.peek_from_stream(self.stream).frame_index
+            try:
+                if step_size_confidence == -1:
+                    position = self.assumptuous_seek_position_(frame_index)
+                else:
+                    position = self.assumptuous_seek_position_(
+                        min(current_frame_index + step_size_confidence, frame_index)
+                    )
+                    step_size_confidence *= 10
+            except Exception:
+                step_size_confidence = 1
 
-    def empty(self):
-        return self.stream.peek(1) == b""
+            if position is not None and position.header.frame_index == frame_index:
+                break
+
+            attempts += 1
 
-    def assumptuous_seek_(self, frame_index: types.FRAME_INDEX):
-        self.stream.seek(self.closest_stored_position(frame_index))
+            if attempts >= 50:
+                break
+
+        if position is None or position.header.frame_index != frame_index:
```

Comment:
Correct me if I'm wrong but it seems we never brute force seek anymore? Couldn't we do that here?

Reply:
Step size = 1 is brute force but it then tries to increase that again

Reply:
Maybe I'm confused but it seems to me that if `assumptuous_seek_position_` repeatedly fails and raises an exception then it gets stuck in the for-loop and `position` never gets set, so it just raises `IndexError`?